File: 859.亲密字符串.py
#
# @lc app=leetcode.cn id=859 lang=python3
#
# [859] 亲密字符串
#

# @lc code=start
class Solution:
    def buddyStrings(self, A: str, B: str) -> bool:
        index =[]
        if len(A) == len(B):
            for i in range(len(A)):
                if A[i] != B[i]:
                    index.append(i)
            if len(index) == 2 and A[index[0]] == B[index[1]] and A[index[1]] == B[index[0]]:
                return True
            if len(index) == 0 and len(A)-len(set(A)) > 0:
                return True
        return False
# @lc code=end

File: 859.亲密字符串.cpp
/*
 * @lc app=leetcode.cn id=859 lang=cpp
 *
 * [859] 亲密字符串
 */

// @lc code=start
class Solution {
public:
    bool buddyStrings(string A, string B) {
        if(A.size() != B.size() or A.empty())
            return false;
        bool flag = false;
        int pos = 0;
        unordered_map<char, int>com;
        if(A == B){
            for(auto ch : A)
                com[ch] += 1;
            for(auto item : com)
                if(item.second > 1)
                    return true;
            return false;
        }
        for(int i = 0; i < A.size(); i++){
            if(A[i] != B[i]){
                if(flag)
                {
                    swap(A[i], A[pos]);
                    return A == B;
                }
                flag = true;
                pos = i;
            }
        }
        return false;
    }
};
// @lc code=end

